### DISCUSSION TITLE
fix(freeze) rows and columns

### DISCUSSION THREAD

**Discussion by SergeyZaz** [QtExcel/QXlsx#200] [Ideas]
Hello! I didn't find a way to freeze rows and columns. I did it myself, please add it to the repository.

in xlsxabstractsheet.h
` 
   QVector<CellLocation> getFullCells(int* maxRow, int* maxCol);
    
	///////////////////////// ZAZ: fix(freeze) rows and columns! /////////////////////////
	void setFrozenRows(int rows);
    void setFrozenColumns(int cols);

private:
`

in xlsxabstractsheet_p.h
`
    int outline_col_level;

	///////////////////////// ZAZ: fix(freeze) rows and columns! /////////////////////////
	int frozen_rows;
    int frozen_cols;

    int default_row_height;
`

in xlsxworksheet.cpp
`
	default_row_zeroed = false;
	
	///////////////////////// ZAZ: fix(freeze) rows and columns! /////////////////////////
	frozen_rows = 0;
	frozen_cols = 0;
}
`
`
writer.writeAttribute(QStringLiteral("workbookViewId"), QStringLiteral("0"));

	///////////////////////// ZAZ: fix(freeze) rows and columns! /////////////////////////
	if (d->frozen_cols > 0 || d->frozen_rows > 0)
	{
		writer.writeStartElement(QStringLiteral("pane"));
		if (d->frozen_cols > 0)
			writer.writeAttribute(QStringLiteral("xSplit"), QString::number(d->frozen_cols));//задается количество фиксированных столбцов
		if (d->frozen_rows > 0)
			writer.writeAttribute(QStringLiteral("ySplit"), QString::number(d->frozen_rows));//задается количество фиксированных строк
		writer.writeAttribute(QStringLiteral("topLeftCell"), CellReference(d->frozen_rows + 1, d->frozen_cols + 1).toString());// надо задать следующую за зафиксированной областью ячейку
		writer.writeAttribute(QStringLiteral("activePane"), QStringLiteral("bottomRight"));
		writer.writeAttribute(QStringLiteral("state"), QStringLiteral("frozenSplit"));
		writer.writeEndElement();//pane
	}
	
	writer.writeEndElement();//sheetView
`
`
//{{ liufeijin 
bool Worksheet::setStartPage(int spagen)
{
    Q_D(Worksheet);

    d->PfirstPageNumber=QString::number(spagen);

    return true;
}
//}}

///////////////////////// ZAZ: fix(freeze) rows and columns! /////////////////////////
void Worksheet::setFrozenRows(int rows)
{
    Q_D(Worksheet);

	d->frozen_rows=rows;
}

void Worksheet::setFrozenColumns(int cols)
{
   Q_D(Worksheet);

   d->frozen_cols=cols;
}

`

**Comment by j2doll** [comment]
Dear @SergeyZaz 

Sorry for late reply.

Thank you for your good opinion.

I will make new branch for your opinion(later). 😃

**Comment by AndreiCherniaev** [comment]
Fixed by [commit](https://github.com/QtExcel/QXlsx/pull/352/commits/230af505fedc40af3119f63e0645269853404510)
